### DISCUSSION TITLE
Enterprise SSDs

### DISCUSSION THREAD

**Discussion by WillBVT** [smartbitcoin/arc_plot#14] [Q&A] [answered]
Hi Everyone,

So I've been having a hard time running Arc_plot consistently with my current setup this is both CPU and GPU plotting. CPU plotting is okay but typically crashes in phase two after 10 - 20 plots. GPU plotting almost always crashes at the start of phase 2. I suspect the issue is SSD stability/speed as I'm not able to detect and CPU or RAM instability issues on my system, and other plotters seem to work fine. 

My current setup:
CPU: R9 5900
RAM: 128GB DDR4 3600
Mobo: Asus X570 TUF 
NVME1: Samsung 980 Pro 1TB
NVME2: Samsung 980 Pro 1TB
GPU: Nvidia 3070TI

So my questions are as follows. 
If I wanted to upgrade my SSDs what would people suggest? My thought is to go with one or two PCIe 4 2.5" enterprise u.2 SSDs. My motherboard has two PCIe 4 NVME m.2 slots which I can convert to u.2 with an adapter card. What brand/model would people recommend that has a reasonable price on the used market? Let's say $400 USD

I've noticed that most enterprise SSD's have lower sequential read/write and IOPS compared to the 980 Pros I currently use. Do people find this is actually the case in real world use? If so it seems like it might cause more issues going to SSDs with lower throughput. 

Thanks in advance for any responses.

**Comment by smartbitcoin** [comment, marked as answer]
your SSD should work. just create btrfs raid0 with it and using following script. but keep in mind the copy out speed from ssd to hdd need great than the plot create speed in GPU with certain margin , eg: if gpu plot take 10 mins,   copy speed from ssd to hdd should be less than 9 min.  And after every plot was done , call fstrim to recycle your ssd raid0.

create raid0
if your 980 pro is nvme0n1 & nvme1n1
```
sudo mkfs.btrfs -n 64k -d raid0 -m raid0 /dev/nvme0n1 /dev/nvme1n1 -f
mkdir /ssd
mount -o nodatacow,ssd,noatime,discard,nodatasum,nobarrier,autodefrag /dev/nvme0n1 /ssd
sudo chown -R your_linux_id  /ssd
mkdir /ssd/tmp /ssd/plot
```
then use /ssd/tmp/  and /ssd/plot/ as ssd_tmp_dir and ssd_plot_dir in following script. let's assume your hdd mounted into /mnt/tmp/



plot.sh  ( make sure you in sudo group and can call sudo fstrim without pwd as it used in this script.)
```
#!/bin/sh
# your pool key
pool_key=
# your farm key
farm_key=
# your cpu core number
cpus=12

# your SSD tmp drive
tmp_dir=/ssd/tmp/
# your SSD plot drive
ssd_plot_dir=/ssd/plot/
#your HDD final plot drive
hdd_plot_dir=/mnt/tmp/

# how many plot your wana create.
cnt=80

for(( c=1; c<$cnt; c++))
do
	echo "start plot " ${c}
	./arc_plot -G --gthreads 4 -n 1 -u 256 -v 256 -r $cpus -t $tmp_dir -d $ssd_plot_dir -c $pool_key -f $farm_key 1>log 2>&1
	sleep 3
	rm ${tmp_dir}*
	sleep 3
	plotname=`cat log | grep "plot-"`
	echo "move out plot " ${ssd_plot_dir}${plotname} " into hdd"
	mv ${ssd_plot_dir}${plotname}.plot ${hdd_plot_dir} &
	sleep 3
	echo "fstrim SSD tmp drive"
	sudo fstrim ${tmp_dir}
	sleep 3
done
```

**Reply by WillBVT** [reply]
Hi, 
Thanks a lot for the response. I made both SSDs into a btrfs raid0 array as suggested, and made the bash file similar to your example. 
However, now it hangs during phase 1, all utilization drops to zero on CPU and GPU. I've let it sit this way for maybe 20 minutes but no error comes.

[log.txt](https://github.com/smartbitcoin/arc_plot/files/10333683/log.txt)

Any ideas?

**Reply by smartbitcoin** [reply]
cpu & gpu no usage means it's waiting for IO.  check your ssd health status with this cmd.
``` smartctl -a /dev/nvme0n1```

and 
``` sudo dmesg``` to check if any IO error in it.

**Comment by WillBVT** [comment]
Hey,

So I checked the smart status of both nvme drives and they both look good. However, when I ran the dmesg command I see a bunch of "out of range address" errors on the GPU. I attached an excerpt of the log file below.

Also I've noticed that if I try to run with less than 512 buckets when gpu plotting it crashes my system, and I have to do a hard reboot? Do you think the system is running out of memory? 

[demsg.txt](https://github.com/smartbitcoin/arc_plot/files/10334130/demsg.txt)

**Reply by smartbitcoin** [reply]
use 
```nvidia-smi```
to check your nvidia driver version,  should > 450.xx

use
```nvtop```
to monitor your gpu vram usage, make sure you have proper configure so it won't run out of vram.

```arc_plot -G --gthreads 3```  should be good for 8G vram gpu , if you set ```--gthreads ``` too big, it will run out of vram .  try ```--gthreads 1```as  safe mode to confirm gpu works first before increase that value .

if gpu stuck, you need reboot or reload gpu driver. and if you use ubunut desktop , make sure no other application use GPU too much. use only 256 for all buckets for GPU plot.

**Comment by WillBVT** [comment]
Hi again,

Thanks for the suggestions about the different GPU threads. I have had a lot of success plotting using 1 thread.  Using more than 1 thread seems to crash the GPU driver after P1 table 1. 2 threads sometimes can complete 1 plot but typically crashes during the second pot. 3 threads crashes every time. I've monitored in Nvtop and I'm definitely not running out of GPU memory.
